### PR TITLE
Adding replace_where incremental strategy

### PR DESCRIPTION
**Pull request by andrefurlan-db**
Resolves #

### Description
This PR adds a new incremental strategy `replace_where`. The strategy resolves to an `INSERT INTO ... REPLACE WHERE` statement. This completes the feature set explained here: https://docs.databricks.com/delta/selective-overwrite.html#replace-where&language-python

A lot of the code change is to bring macros from dbt-spark over to dbt-databricks. The only real code change was in validating incremental strategies and adding the replace_where strategy.

#### Why do we need it?
It enables use cases where part of the data is always replaced and where MERGE is not possible, such as when there is no primary key. E.g.: events table where we want to always replace the last 3 days.

#### Difference from insert_overwrite
Insert overwrite only works with dynamic partition pruning spark setting, which is not available in sql warehouses or any Unity Catalog-enabled cluster. It also only works with whole partitions, making it difficult to set up and assure that the correct data is dropped.

 

### Checklist

- [x] I have run this code in development and it appears to resolve the stated issue
- [x] This PR includes tests, or tests are not required/relevant for this PR
- [ ] I have updated the `CHANGELOG.md` and added information about my change to the "dbt-databricks next" section.

### CODE REVIEW

File: tests/integration/incremental_strategies/test_incremental_strategies.py [hunk 1]
```diff
@@ -58,29 +58,40 @@ class TestDeltaStrategies(TestIncrementalStrategies):
...
-    def run_and_test(self):
+    def run_and_test_cluster(self):
+        self.seed_and_run_twice()
+        self.assertTablesEqual("append_delta", "expected_append")
+        self.assertTablesEqual("merge_no_key", "expected_append")
+        self.assertTablesEqual("merge_unique_key", "expected_upsert")
+        self.assertTablesEqual("merge_update_columns", "expected_partial_upsert")
+        self.assertTablesEqual("merge_exclude_columns", "expected_exclude_upsert")
+        self.assertTablesEqual("insert_overwrite_no_partitions", "expected_overwrite")
+        self.assertTablesEqual("insert_overwrite_partitions", "expected_upsert")
```
Comment (ueshin) [inline tests/integration/incremental_strategies/test_incremental_strategies.py:69]: Seems like the models are missing?

File: dbt/include/databricks/macros/materializations/incremental/strategies.sql [hunk 2]
```diff
@@ -6,6 +6,89 @@
...
+
+    {%- set dest_columns = adapter.get_columns_in_relation(target_relation) -%}
+    {%- set dest_cols_csv = dest_columns | map(attribute='quoted') | join(', ') -%}
+    insert overwrite table {{ target_relation }}
+    {{ partition_cols(label="partition") }}
+    select {{dest_cols_csv}} from {{ source_relation }}
+
+{% endmacro %}
+
+{% macro get_replace_where_sql(args_dict) -%}
```
[the remark this answers is not in the document]
Comment (andrefurlan-db) [reply dbt/include/databricks/macros/materializations/incremental/strategies.sql:28]: no? Let me double check

File: tests/integration/incremental_strategies/test_incremental_strategies.py (hunk 1 above)
Comment (andrefurlan-db) [reply tests/integration/incremental_strategies/test_incremental_strategies.py]: good catch! fixed in next commit

File: dbt/include/databricks/macros/materializations/incremental/strategies.sql (hunk 2 above)
[the remark this answers is not in the document]
Comment (andrefurlan-db) [reply dbt/include/databricks/macros/materializations/incremental/strategies.sql:34]: good catch. Fixed in next commit

**Review by ueshin — APPROVED** [review]

**Pull request merged**